crates/accelerate/src/commutation_cancellation.rs

```diff
@@ -0,0 +1,291 @@
+    second_index: Option<usize>,
+}
+
+#[pyfunction]
+#[pyo3(signature = (dag, commutation_checker, basis_gates=None))]
+pub(crate) fn cancel_commutations(
+    py: Python,
+    dag: &mut DAGCircuit,
+    commutation_checker: &mut CommutationChecker,
+    basis_gates: Option<HashSet<String>>,
+) -> PyResult<()> {
+    let basis: HashSet<String> = if let Some(basis) = basis_gates {
+        basis
+    } else {
+        HashSet::new()
+    };
+    let z_var_gate = dag
+        .op_names
+        .keys()
+        .find_map(|g| {
+            VAR_Z_MAP
+                .iter()
+                .find(|(key, _)| *key == g.as_str())
+                .map(|(_, gate)| gate)
+        })
+        .or_else(|| {
+            basis.iter().find_map(|g| {
+                VAR_Z_MAP
+                    .iter()
+                    .find(|(key, _)| *key == g.as_str())
+                    .map(|(_, gate)| gate)
+            })
+        });
+    // Fallback to the first matching key from basis if there is no match in dag.op_names
+
+    // Gate sets to be cancelled
+    /* Traverse each qubit to generate the cancel dictionaries
+     Cancel dictionaries:
+      - For 1-qubit gates the key is (gate_type, qubit_id, commutation_set_id),
+        the value is the list of gates that share the same gate type, qubit, commutation set.
+      - For 2qbit gates the key: (gate_type, first_qbit, sec_qbit, first commutation_set_id,
+        sec_commutation_set_id), the value is the list gates that share the same gate type,
+        qubits and commutation sets.
+    */
+    let (commutation_set, node_indices) = analyze_commutations_inner(py, dag, commutation_checker)?;
+    let mut cancellation_sets: HashMap<CancellationSetKey, Vec<NodeIndex>> = HashMap::new();
+
+    (0..dag.num_qubits() as u32).for_each(|qubit| {
+        let wire = Qubit(qubit);
+        if let Some(wire_commutation_set) = commutation_set.get(&Wire::Qubit(wire)) {
+            wire_commutation_set
+                .iter()
+                .enumerate()
+                .for_each(|(com_set_idx, com_set)| {
+                    // This ensures that we only have DAGOPNodes in the current com_set, yuck...
+                    if let NodeType::Operation(_node0) = &dag.dag[*com_set.first().unwrap()] {
+                        com_set.iter().for_each(|node| {
+                            let op = match &dag.dag[*node] {
+                                NodeType::Operation(instr) => instr,
+                                _ => panic!("Unexpected type in commutation set."),
+                            };
+                            let num_qargs = dag.get_qargs(op.qubits).len();
+                            // no support for cancellation of parameterized gates
+                            if op
+                                .params_view()
+                                .iter()
+                                .all(|p| !matches!(p, Param::ParameterExpression(_)))
+                            {
```

Comment:
There is `op.is_parameterized()` here which you can probably use here instead with:

```suggestion
                            if !op.is_parameterized() {
```